include/openmc/event.h

```diff
@@ -0,0 +1,108 @@
+#ifndef OPENMC_EVENT_H
+#define OPENMC_EVENT_H
+
+//! \file event.h
+//! \brief Event-based data structures and methods
+
+#include "openmc/particle.h"
+#include "openmc/shared_array.h"
+
+
+namespace openmc {
+
+//==============================================================================
+// Structs
+//==============================================================================
+
+// In the event-based model, instead of moving or sorting the particles
+// themselves based on which event they need, a queue is used to store the
+// index (and other useful info) for each event type.
+// The QueueItem struct holds the relevant information about a particle needed
+// for sorting the queue. For very high particle counts, a sorted queue has the
+// potential to result in greatly improved cache efficiency. However, sorting
+// will introduce some overhead due to the sorting process itself, and may not
+// result in any benefits if not enough particles are present for them to achieve
+// consistent locality improvements. 
+struct QueueItem{
+  int64_t idx;         //!< particle index in event-based particle buffer
+  Particle::Type type; //!< particle type
+  int64_t material;    //!< material that particle is in
+  double E;            //!< particle energy
+
+  // Compare by particle type, then by material type, then by energy
+  // TODO: Currently, material IDs are not usually unique to material
+  // types. When unique material type IDs are available, we can alter the
+  // material field in this struct to contain the material type instead.
```

Comment:
In a typical reactor, we may only have a few different types of fuel materials. The use of "type" of material here is in terms of which isotopes are present in that material, but not their densities. A typical reactor may only have a handful of different fuel material "types". In depleted fuel, each "instance" of a fuel pin would have its own specific isotopic densities. A typical reactor would therefore have tens of thousands or more different material instances. In OpenMC, currently material IDs correspond to specific instances rather than more general types.

Ideally, we would like to sort by "type" of material rather than "instance". This would allow contiguous lookup items to be cycling through the exact same list of isotopes, while still applying different isotopic densities to the lookups. Currently, we are sorting by instance, which means that adjacent lookups may or may not tend to be operating on the same list of nuclides.

@paulromano had indicated that in the near future we may have an additional ID present that indicates the "type" of material instead of just the "instance" of the material, in which case we would want to use this new "type" index when we enqueue an element.

Reply:
To be a little more specific, what I am thinking about is to have sort of a "shared composition" for cases where you have many materials, all of which have the same nuclides. In this case, we currently store the list of nuclides for each of the material; if we were to instead have a class that stored the nuclide indices separate from the densities, we could cut down on the memory requirement a bit. I mentioned to @jtramm that this would make for an effective way to sort events by material type (in the case that all fuel materials shared the nuclide list through this idea).

For now, it might be good to just have some language that indicates what you're getting at, like "Compare by particle type, then by material type (e.g., fuel/nonfuel)..."

Reply:
I added in a bit more discussion in the comment here to hopefully make it more clear, but let me know if you had specific verbiage in mind for this.